### DISCUSSION TITLE
default pins for SPI, I²C, Serial

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#1819] [Q&A] [answered]
In the wiki, for Hardwareserial it says that when creating a new serial instance, the first pin in pinmap corresponding to that serial port is chosen.
Does this also apply to additional SPI and I²C instances, or is it mandatory to specify them upon creation?
Where are the pins for the default SPI and I²C defined? I checked for Nucleo-L4R5 and the "first found" rule does not apply to I²C SDA and SCLK, and also not for SPI SCLK and NSS (which uses a pin not defined in the corresponding pinmap).
I have not been able to find where these pins are defined. Neither in variant_NUCLEO-L4R5.h nor in Peripherals.h

additional question: If I wanted to create a custom variant in which more serials and second I²C and SPI are defined by default, where (and how) can I add those? In variant_XXX.h?

**Comment by fpistm** [comment, marked as answer]
> In the wiki, for Hardwareserial it says that when creating a new serial instance, the first pin in pinmap corresponding to that serial port is chosen.

Yes if you only provide the peripheral instance name.


> Does this also apply to additional SPI and I²C instances, or is it mandatory to specify them upon creation? 

Yes, same if you specify instance name.

> Where are the pins for the default SPI and I²C defined? 

If they are not defined in the` variant_*.h` then it uses the default one which are specified here:

https://github.com/stm32duino/Arduino_Core_STM32/blob/main/cores/arduino/pins_arduino.h#L43-L84

> I checked for Nucleo-L4R5 and the "first found" rule does not apply to I²C SDA and SCLK, and also not for SPI SCLK and NSS (which uses a pin not defined in the corresponding pinmap). I have not been able to find where these pins are defined. Neither in variant_NUCLEO-L4R5.h nor in Peripherals.h

As it is a Nucleo with an Arduino Uno connector, it uses the default.  For NSS it could be any pin not only the hardware one from the peripheral as in that case it is a Software CS.
 
> additional question: If I wanted to create a custom variant in which more serials and second I²C and SPI are defined by default, where (and how) can I add those? In variant_XXX.h?

Several way exists:
 - Define your own board in the core: https://github.com/stm32duino/wiki/wiki/Add-a-new-variant-%28board%29
 - Use the core and define new variant: https://github.com/stm32duino/wiki/wiki/Custom-board-based-on-a-core
